lib/charms/sdcore_nrf/v0/fiveg_nrf.py

```diff
@@ -0,0 +1,281 @@
+        """Handler triggered on relation changed event.
+
+        Args:
+            event: Juju event (RelationChangedEvent)
+
+        Returns:
+            None
+        """
+        if remote_app_relation_data := self._get_remote_app_relation_data(event.relation):
+            self.on.nrf_available.emit(url=remote_app_relation_data["url"])
+
+    def get_nrf_url(self) -> Optional[str]:
+        """Returns NRF url.
+
+        Returns:
+            str: NRF url.
+        """
+        if remote_app_relation_data := self._get_remote_app_relation_data():
+            return remote_app_relation_data.get("url")
+        return None
+
+    def _get_remote_app_relation_data(
+        self, relation: Optional[Relation] = None
+    ) -> Optional[Dict[str, str]]:
+        """Get relation data for the remote application.
+
+        Args:
+            Relation: Juju relation object (optional).
+
+        Returns:
+            Dict: Relation data for the remote application
+            or None if the relation data is invalid.
+        """
+        relation = relation or self.model.get_relation(self.relation_name)
+        if not relation:
+            logger.warning(f"No relation: {self.relation_name}")
+            return None
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relation_name)
+            return None
+        remote_app_relation_data = dict(relation.data[relation.app])
+        if not self._relation_data_is_valid(remote_app_relation_data):
+            logger.error("Invalid relation data: %s", remote_app_relation_data)
+            return None
+        return remote_app_relation_data
+
+    @staticmethod
+    def _relation_data_is_valid(relation_data: dict) -> bool:
+        """Returns whether URL is valid.
+
+        Args:
+            str: URL to be validated.
+
+        Returns:
+            bool: True if URL is valid, False otherwise.
+        """
+        try:
+            ProviderSchema(app=relation_data)
+            return True
+        except ValidationError:
+            return False
+
+
+class NRFProvides(Object):
+    """Class to be instantiated by the charm providing the NRF data."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+        self.charm = charm
+
+    @staticmethod
+    def _relation_data_is_valid(url: str) -> bool:
+        """Returns whether URL is valid.
+
+        Args:
+            str: URL to be validated.
+
+        Returns:
+            bool: True if URL is valid, False otherwise.
+        """
+        try:
+            ProviderSchema(app=ProviderAppData(url=url))  # type: ignore[arg-type]
+            return True
+        except ValidationError as e:
+            logger.error("Invalid url: %s", e)
+            return False
+
+    def set_nrf_information(self, url: str) -> None:
```

Comment:
A more pythonic way of doing this would be with a property setter called `nrf_url`:

```suggestion
    @property
    def nrf_url(self) -> str:
        ...
        
    @nrf_url.setter    
    def set_nrf_information(self, url: str) -> None:
        ...
```

Then, it can be used in code like this:

```
    self.nrf_url = "https://my_awesome_nrf.com"
```

Reply:
This is cool, however, following Guillaume's comment, I've added a new parameter to the setter `set_nrf_information(self, url: str, relation_name: Optional[str] = None)`, I don't think it's possible now.

```python
self.nrf_url = "https://my_awesome_nrf.com", "awesome-relation" # <--- This doesn't seem right
```

Let me know if there's any other way, cool trick honestly.

Reply:
Unfortunately, to have a property setter, we need to also have the initial property. It definitely helps in making APIs that are more Pythonic, and since this is for a public API, I think it might be worth it. Anytime we have getter and setter methods, it is a bit of a smell.

It is not super critical however. But if we decide to do it, it needs to be done now, otherwise it will be a breaking change down the line.